packages/components/modal/src/Modal.tsx

```diff
@@ -143,19 +143,32 @@ export const Modal = ({
     overlayClassName: otherProps.overlayProps?.className,
   });
 
-  React.useEffect(() => {
-    if (props.isShown) {
-      setTimeout(() => {
-        if (props.initialFocusRef && props.initialFocusRef.current) {
-          if (props.initialFocusRef.current.focus) {
-            props.initialFocusRef.current.focus();
-          }
-        } else if (contentRef.current) {
-          focusFirstWithinNode(contentRef.current);
-        }
-      }, 100);
+  const onInitialFocus = () => {
+    const contentEl = contentRef.current;
+    if (contentEl) {
+      const activeEl = document.activeElement;
+      const isContentContainsActive =
+        contentEl !== activeEl && contentEl.contains(activeEl);
+
+      if (isContentContainsActive) {
+        return;
+      }
+    }
+
+    const initialFocusEl = props.initialFocusRef?.current;
+    if (initialFocusEl) {
+      initialFocusEl.focus?.();
+    } else if (contentEl) {
+      focusFirstWithinNode(contentEl);
     }
-  }, [props.isShown, props.initialFocusRef]);
+  };
+
+  const onAfterOpen: ModalProps['onAfterOpen'] = (...args) => {
```

Comment:
We should probably wrap in `useCallback` here so that the value passed to the modal is stable between renders

Reply:
i have to disagree, in order to make it _work_, we have to either force users to wrap `onAfterOpen` prop in `useCallback`, or save the original function to a ref object in order to avoid stale closure problem.

but all this doesn't make much sense, because `react-modal` doesn't rely on `onAfterOpen` function identity
https://github.com/reactjs/react-modal/blob/master/src/components/ModalPortal.js#LL229-L253C5

Reply:
If it's not a dependency in react-modal, then makes sense to me 👍